### ISSUE TITLE
Consistently handle errors within toolbox

### ISSUE THREAD

**Issue by PerilousApricot** [accre/lstore#79]
The current toolbox code handles errors non-uniformly and, worse, really poorly in places. Go through and ensure:
- Return codes are uniform, e.g.
  - Functions returning a pointer are NULL on error, pointer otherwise
  - Functions returning integral types are zero on success, otherwise errno (in the linux <errno.h>sense, though we can define our own if we really need to distinguish further)
- Assertions are for things that **can't** happen, not things that **shouldn't** happen
- Don't bomb on errors. Clean up to a sane state and return an error uphill. Unwind with goto error handlers. Goto counters increase through the main body, then reverse in the error handling:

``` c
int x() {
  if (!foo())
    goto error_1;
  if (!bar())
    goto error_2;
  return 0;
  error_2:
    cleanup_bar();
  error_1:
    cleanup_foo();
  return 1;
}
```
- Bombing kills any chance to log context of the error
  - Basically need a debug build + debug symbols + gdb to trace down why the process just stopped working (a current problem with the ibp server w/DLT)
- Bombing is opaque to the caller.
  - Someone developing against the library will just see `atexit()` fire
